python-lib/design/ab_dispatcher.py

```diff
@@ -0,0 +1,106 @@
+        self.size_A = size_A
+        self.size_B = size_B
+        self.reference_column = reference_column
+
+    def dispatch(self, experiment_population: np.array, attribution: AttributionMethod, input_df: pd.DataFrame) -> tuple:
+        """Dispatch the experiment population into two groups
+
+        :param np.array experiment_population: array containing references to the experiment population
+        :param AttributionMethod attribution: defines how to deal with the leftover users.
+        :raises: :class:`ValueError`: Not enough users to run the experiment
+
+        :returns: Experiment dataset and two arrays containing the contacts for each group
+        :rtype: tuple
+        """
+        self.check_sample_size(experiment_population)
+        A_group, B_group = self.split_into_groups(experiment_population, attribution)
+        group_df = self.add_group_column(input_df, A_group, B_group)
+        logger.info("Dispatching experiment population: done")
+        return group_df, A_group, B_group
+
+    def check_sample_size(self, experiment_population: np.array):
+        """Check if the size of the population is large enough to run the experiment.
+
+        :raises: :class:`ValueError`: Not enough users to run the experiment
+        """
+        population_size = experiment_population.shape[0]
+
+        logger.info("Dispatching experiment population ...")
+        logger.info("Dispatching experiment population: checking the sample sizes")
+        logger.info("Sample size of the A group:" + str(self.size_A))
+        logger.info("Sample size of the B group:" + str(self.size_B))
+        logger.info("Size of the experiment population: " + str(population_size))
+
+        if self.size_A + self.size_B > population_size:
+            raise ValueError("Not enough user ids or emails to run the current AB testing.")
+
+    def split_into_groups(self, experiment_population: np.array, attribution_method: str) -> tuple:
+        """Shuffle the population and splits it into groups
+
+        :returns: Two arrays containing the contacts for each group
+        :rtype: tuple
+        """
+        logger.info("Dispatching experiment population: shuffle")
+        np.random.permutation(experiment_population)
+        if attribution_method == AttributionMethod.LEFTOVER_TO_A:
+            B_group = experiment_population[:self.size_B]
+            A_group = experiment_population[self.size_B:]
+        elif attribution_method == AttributionMethod.LEFTOVER_TO_B:
+            A_group = experiment_population[:self.size_A]
+            B_group = experiment_population[self.size_A:]
+        elif attribution_method == AttributionMethod.LEFTOVER_BLANK:
+            B_group = experiment_population[:self.size_B]
+            A_group = experiment_population[self.size_B:self.size_A+self.size_B]
+        return A_group, B_group
+
+    def add_group_column(self, input_df: pd.DataFrame, A_group: np.array, B_group: np.array) -> pd.DataFrame:
```

Comment:
It is usually a bad idea to directly modify and return the parameter of your function. You can check out this thread: https://stackoverflow.com/questions/575196/why-can-a-function-modify-some-arguments-as-perceived-by-the-caller-but-not-oth
and this https://www.dataquest.io/blog/tutorial-functions-modify-lists-dictionaries-python/

Reply:
done in https://github.com/dataiku/dss-plugin-ab-testing/pull/1/commits/42e6adc2eb62d94294976c37afbd072cbbf53120